File: Searching/binarysearch.cpp
#include<bits/stdc++.h>
using namespace std;

int binarysearchrec(int *arr,int l,int h,int k){
    int mid=(l+h)/2;
    if(arr[mid]==k)
        return mid;
    if(arr[mid]<k)
        return binarysearchrec(arr,mid+1,h,k);


  return binarysearchrec(arr,l,mid,k);        

}

int binarysearch(int *arr,int l,int h,int k){
    int mid;
  
    while(l<=h){
        mid=l+(h-l)/2;
         if(arr[mid]==k)
            return mid;
        if(arr[mid]>k)
            h=mid;
        else if(arr[mid]<k)
            l=mid+1;    
    } 
   
    return -1;

}
int main(){
    int n;
    cout<<"Enter size of array "<<endl;
    cin>>n;
    int* arr=new int(n);
    cout<<"Enter Array Elements "<<endl;
    for(int i=0;i<n;i++)
        cin>>arr[i];
    int k;
    cout<<"Enter element to be searched"<<endl;
    cin>>k;
    int l=0;
    int h=n-1;
    cout<<"Element found at index : "<<binarysearch(arr,l,h,k)<<endl;
    cout<<"Element found at index (using recursion) : "<<binarysearchrec(arr,l,h,k)<<endl;
}

File: Searching/binarysearch.py

def binarysearchrec(arr,l,h,k):
    if(l<=h):
        mid=(l+h)//2
        if arr[mid]==k:
            return mid
        if arr[mid]>k:
            binarysearchrec(arr,mid+1,h,k)
        if arr[mid]<k:
            binarysearchrec(arr,l,mid-1,k)



def binarysearchit(arr,l,h,k):
    if(l<h):
        while l<h:
            mid=(l+h)//2
            if arr[mid]==k:
                return mid
            elif arr[mid]>k:
                h=mid-1
            else:
                l=mid+1    
                    

arr =[10,20,30,40,50]
l=0
h=4
k=30
ans=binarysearchrec(arr,l,h,k)
print("Element is Present at index (using recursion) ",ans)
ans2=binarysearchit(arr,l,h,k)
print("Element is Present at index (iterative method)",ans2)